runtime/service.go

```diff
@@ -436,9 +361,296 @@ func (s *service) nextVSockPort() uint32 {
 	return port
 }
 
+func (s *service) waitVMReady() error {
+	select {
+	case <-s.vmReady:
+		return nil
+	case <-time.After(firecrackerStartTimeout):
+		return status.Error(codes.DeadlineExceeded, "timed out waiting for VM start")
+	}
+}
+
+func (s *service) CreateVM(requestCtx context.Context, request *proto.CreateVMRequest) (*empty.Empty, error) {
+	errCh := make(chan error, 1)
+
+	s.vmStartOnce.Do(func() {
```

Comment:
Can we break this into multiple functions? This function is a little long, so I think having things like `createMachine` will at least group things together nicely.

Reply:
Yeah at min I can extract out everything in the `Once.Do`, I'll also see if breaking that function up some more is beneficial.